### DISCUSSION TITLE
PM2 Cluster Mode Bug when adding Fields

### DISCUSSION THREAD

**Discussion by Michael-Nussbaumer** [directus/directus#16381] [Archive]
### Describe the Bug

I am encountering this bug on 2 installations right now, they are both running in PM2 Cluster Mode. If I uses more as one instance, I get an 403 Error when adding new Fields to a collection. So there seems to be an issue between instances with the authentication. 

Everything else seems to work fine.

Reducing the clusters to 1 instance solves this issue. 

We were alread thinking about an extra "Admin" installation as single instance and have the API running as a cluster.

### To Reproduce

Setup Directus with pm2 in cluster mode and try to add fields to a collection. (I assume, the more instances you have, the more likely you get the bug)

### Errors Shown

403

### What version of Directus are you using?

9.20.4

### What version of Node.js are you using?

16.17.0

### What database are you using?

MySQL 8

### What browser are you using?

Chrome

### How are you deploying Directus?

linux vps

**Comment by azrikahar** [comment]
By `403`, is it "[FORBIDDEN] You don't have permission to access this." or something else?

Are you using redis for the cache? When running directus in clusters (pm2, kubernetes etc), it would be important to make sure not to default to the `memory` cache as `memory` would not work properly in such scenarios.

**Comment by Michael-Nussbaumer** [comment]
> By `403`, is it "[FORBIDDEN] You don't have permission to access this." or something else?
> 
> Are you using redis for the cache? When running directus in clusters (pm2, kubernetes etc), it would be important to make sure not to default to the `memory` cache as `memory` would not work properly in such scenarios.

Yes, but I am on the admin account, so 403 shouldn't be the case. 

Currently the system is in development, so `CACHE_ENABLED=false`

**Comment by azrikahar** [comment]
> Yes, but I am on the admin account, so 403 shouldn't be the case.

Technically yes, but currently the API does still returns 403 errors when something does not exist rather than "insufficient permission", so in that case, an admin account will still receive 403 errors (my current suspicion is on schema that is not in sync, i.e. one instance has the created field, but the other doesn't, so the other throws an error). Understandably that can be rather confusing, and it is being discussed in #4368.

> Currently the system is in development, so `CACHE_ENABLED=false`

Have you disabled `CACHE_SCHEMA` and `CACHE_PERMISSIONS` as well? If not, can you try disabling them? Note that they are independent of `CACHE_ENABLED`, so they can still be `true` i.e. enabled. Additional reference to [this notice in the docs](https://docs.directus.io/self-hosted/config-options.html#cache) for more info:

![](https://user-images.githubusercontent.com/42867097/200238213-869c3ce9-e233-4548-a6ae-cc7801c1f26f.png)

**Comment by Michael-Nussbaumer** [comment]
If I disable `CACHE_SCHEMA` and `CACHE_PERMISSIONS` the problem is gone.

**Reply by azrikahar** [reply]
That confirms that when you were running in cluster mode, they are indeed caching the schema and permission in memory, so each instance will de-sync whenever there are changes such as adding fields 👍

 The recommended approach in [the notice above](https://github.com/directus/directus/discussions/16381#discussioncomment-4098253) would still be the solution in such setup:

> When running Directus load balanced, you'll need to use a shared cache storage (like [Redis](https://docs.directus.io/self-hosted/config-options.html#redis-2) or [Memcache](https://docs.directus.io/self-hosted/config-options.html#memcache-2))
